lib/autoupdate/tools/migration.go

```diff
@@ -0,0 +1,144 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tools
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// migrateV1AndUpdateConfig launches migration process and add migrated
+// tools to configuration file.
+func migrateV1AndUpdateConfig(toolsDir string, tools []string) error {
+	migratedTools, err := migrateV1(toolsDir, tools)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(migratedTools) == 0 {
+		return nil
+	}
+
+	ctc, save, err := newClientToolsConfig(toolsDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, tool := range migratedTools {
+		ctc.AddTool(tool)
+	}
+
+	return trace.Wrap(save())
+}
+
+// migrateV1 verifies the tool binary located in the tool's directory.
+// If it is a symlink, it reads the target location and generates a tool object
+// to be saved in the configuration for backward compatibility.
+// If it is a regular binary, a new package folder should be created,
+// and the binary should be copied to the new location.
+// TODO(vapopov): DELETE in v21.0.0 - the version without caching will no longer be supported.
+func migrateV1(toolsDir string, tools []string) (map[string]Tool, error) {
+	migratedTools := make(map[string]Tool)
+	for _, tool := range tools {
+		path := filepath.Join(toolsDir, tool)
+		info, err := os.Lstat(path)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return nil, trace.Wrap(err, "failed to retrieve information for tool %q", tool)
+		}
+
+		toolVersion, err := CheckToolVersion(path)
+		if trace.IsBadParameter(err) {
+			// If we can't identify toolVersion, it is blocked by EDR software or binary
+			// is damaged we should continue migration process.
+			slog.ErrorContext(context.Background(), "failed to check the toolVersion", "error", err)
+			continue
+		} else if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if info.Mode().Type()&os.ModeSymlink != 0 {
+			fullPath, err := os.Readlink(path)
+			if err != nil {
+				return nil, trace.Wrap(err, "failed to read symlink %q", path)
+			}
+			pkg, relPath := extractPackageName(toolsDir, fullPath)
+			if t, ok := migratedTools[toolVersion]; ok {
+				t.PathMap[tool] = relPath
+			} else {
+				newPkg := fmt.Sprint(uuid.New().String(), updatePackageSuffixV2)
+				if err := utils.RecursiveCopy(filepath.Join(toolsDir, pkg), filepath.Join(toolsDir, newPkg), nil); err != nil {
+					return nil, trace.Wrap(err)
+				}
+				migratedTools[toolVersion] = Tool{
+					Version: toolVersion,
+					PathMap: map[string]string{tool: relPath},
+					Package: newPkg,
+				}
+			}
+		} else {
```

Comment:
Suggest `continue` here, so you can avoid indenting the else block. You could do the same for the two other else blocks (above and below) as well.